File: Recitation4/Recitation4/Recitation4_HTML5/public_html/js/RedBoxManRenderer.js
/* 
 * This JavaScript file provides methods for clearing the
 * canvas and for rendering Red Box Man.
 */

var canvas;
var gc;
var canvasWidth;
var canvasHeight;
var imageLoaded;
var mousePositionRendering;
var redBoxManImage;
var mouseX;
var mouseY;
var imagesRedBoxManLocations;
var shapesRedBoxManLocations;

function Location(initX, initY) {
    this.x = initX;
    this.y = initY;
}

function init() {
    // GET THE CANVAS SO WE CAN USE IT WHEN WE LIKE
    canvas = document.getElementById("red_box_man_canvas");
    gc = canvas.getContext("2d");
        
    // MAKE SURE THE CANVAS DIMENSIONS ARE 1:1 FOR RENDERING
    canvas.width = canvas.offsetWidth;
    canvas.height = canvas.offsetHeight;
    canvasWidth = canvas.width;
    canvasHeight = canvas.height;
    
        // FOR RENDERING TEXT
    gc.font="32pt Arial";

    // LOAD THE RED BOX MAN IMAGE SO WE CAN RENDER
    // IT TO THE CANVAS WHENEVER WE LIKE
    redBoxManImage = new Image();
    redBoxManImage.onload = function() {
	imageLoaded = true;
    }
    redBoxManImage.src = "./images/RedBoxMan.png";
    
    // BY DEFAULT WE'LL START WITH MOUSE POSITION RENDERING ON
    mousePositionRendering = true;
    
    // HERE'S WHERE WE'LL PUT OUR RENDERING COORDINATES
    imagesRedBoxManLocations = new Array();
    shapesRedBoxManLocations = new Array();
}

function processMouseClick(event) {
    updateMousePosition(event);
    var location = new Location(mouseX, mouseY);
    if (event.shiftKey) {
	shapesRedBoxManLocations.push(location);
	render();
    }
    else if (event.ctrlKey) {
	if (imageLoaded) {
	    imagesRedBoxManLocations.push(location);
	    render();
	}
    }
    else {
	clear();
    }
}

function clearCanvas() {
    gc.clearRect(0, 0, canvasWidth, canvasHeight);
}

function clear() {
    shapesRedBoxManLocations = [];
    imagesRedBoxManLocations = [];
    clearCanvas();
}

function updateMousePosition(event) {
    var rect = canvas.getBoundingClientRect();
    mouseX = event.clientX - rect.left;
    mouseY = event.clientY - rect.top;
    render();
}

function renderShapesRedBoxMan(location) {
    var headColor = "#0000FF";
    var outlineColor = "#000000";
    var eyeColor = "#FFFF00";
    var headW = 115;
    var headH = 88;
    var bodyW = 55;
    var bodyH = 25;
    var legW = 30;
    var legH = 10;
    var feetW = 6;
    var feetH = 6;
    var mouthW = 85;
    var mouthH = 10;
    var eyeW = 30;
    var eyeH = 25;
    
    // DRAW HIS RED HEAD
    gc.fillStyle = headColor;
    gc.fillRect(location.x, location.y, headW, headH);
    gc.beginPath();
    gc.strokeStyle = outlineColor;
    gc.lineWidth = 1;
    gc.rect(location.x, location.y, headW, headH);
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+32, location.y+88, bodyW, bodyH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+45, location.y+112, legW, legH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+40, location.y+122, feetW, feetH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+75, location.y+122, feetW, feetH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+15, location.y+60, mouthW, mouthH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = eyeColor;
    gc.fillRect(location.x+15, location.y+15, eyeW, eyeH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = eyeColor;
    gc.fillRect(location.x+75, location.y+15, eyeW, eyeH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+88, location.y+25, feetW, feetH);
    gc.beginPath();
    gc.stroke();
    
    gc.fillStyle = outlineColor;
    gc.fillRect(location.x+30, location.y+25, feetW, feetH);
    gc.beginPath();
    gc.stroke();
    
    // AND THEN DRAW THE REST OF HIM
}

function renderImageRedBoxMan(location) {
    gc.drawImage(redBoxManImage, location.x, location.y);
}

function renderMousePositionInCanvas(event) {
    if (mousePositionRendering) {
	gc.strokeText("(" + mouseX + "," + mouseY + ")", 10, 50);    
    }
}

function toggleMousePositionRendering() {
    mousePositionRendering = !mousePositionRendering;
}

function render() {
    clearCanvas();
    for (var i = 0; i < shapesRedBoxManLocations.length; i++) {
	var location = shapesRedBoxManLocations[i];
	renderShapesRedBoxMan(location);
    }
    for (var j = 0; j < imagesRedBoxManLocations.length; j++) {
	var location = imagesRedBoxManLocations[j];
	renderImageRedBoxMan(location);
    }
    renderMousePositionInCanvas();
}

File: Recitation4/Recitation4/Recitation4_JavaFX/src/recitation4_javafx/RedBoxManRenderer.java
package recitation4_javafx;

import java.util.ArrayList;
import javafx.application.Application;
import javafx.geometry.Point2D;
import javafx.scene.Group;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;
import javafx.scene.paint.Paint;
import javafx.stage.Stage;

/**
 *
 * @author McKillaGorilla
 */
public class RedBoxManRenderer extends Application {
    Canvas canvas;
    GraphicsContext gc;
    ArrayList<Point2D> imagesRedBoxManLocations;
    ArrayList<Point2D> shapesRedBoxManLocations;
    Image redBoxManImage;
    
    @Override
    public void start(Stage primaryStage) {
	// INIT THE DATA MANAGERS
	imagesRedBoxManLocations = new ArrayList();
	shapesRedBoxManLocations = new ArrayList();
	
	// LOAD THE RED BOX MAN IMAGE
	redBoxManImage = new Image("RedBoxMan.png");
	
	// MAKE THE CANVAS
	canvas = new Canvas();
	canvas.setStyle("-fx-background-color: cyan");
	gc = canvas.getGraphicsContext2D();

	// PUT THE CANVAS IN A CONTAINER
	Group root = new Group();
	root.getChildren().add(canvas);
	
	canvas.setOnMouseClicked(e->{
	    if (e.isShiftDown()) {
		shapesRedBoxManLocations.add(new Point2D(e.getX(), e.getY()));
		render();
	    }
	    else if (e.isControlDown()) {
		imagesRedBoxManLocations.add(new Point2D(e.getX(), e.getY()));
		render();
	    }
	    else {
		clear();
	    }
	});
	
	// PUT THE CONTAINER IN A SCENE
	Scene scene = new Scene(root, 800, 600);
	canvas.setWidth(scene.getWidth());
	canvas.setHeight(scene.getHeight());

	// AND START UP THE WINDOW
	primaryStage.setTitle("Red Box Man Renderer");
	primaryStage.setScene(scene);
	primaryStage.show();
    }
    
    public void clearCanvas() {
	gc.clearRect(0, 0, canvas.getWidth(), canvas.getHeight());
    }
    
    public void clear() {
	shapesRedBoxManLocations.clear();
	imagesRedBoxManLocations.clear();
	render();
    }
    
    public void render() {
	clearCanvas();
	for (int i = 0; i < shapesRedBoxManLocations.size(); i++) {
	    renderShapeRedBoxMan(shapesRedBoxManLocations.get(i));
	}
	for (int j = 0; j < imagesRedBoxManLocations.size(); j++) {
	    renderImageRedBoxMan(imagesRedBoxManLocations.get(j));
	}
    }
    
    public void renderShapeRedBoxMan(Point2D location) {
	/*
        String headColor = "#DD0000";
	String outlineColor = "#000000";
	int headW = 115;
	int headH = 88;
    
	// DRAW HIS RED HEAD
        gc.setFill(Paint.valueOf(headColor));
	gc.fillRect(location.getX(), location.getY(), headW, headH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX(), location.getY(), headW, headH);
	gc.stroke();
        */
	gc.drawImage(redBoxManImage, location.getX(), location.getY());
	// AND THEN DRAW THE REST OF HIM
    }
    
    public void renderImageRedBoxMan(Point2D location) {
	//gc.drawImage(redBoxManImage, location.getX(), location.getY());
        String bodyColor = "#000000";
        String headColor = "#0000FF";
	String outlineColor = "#000000";
        String eyeColor = "#FFFF00";
	int headW = 115;
	int headH = 88;
        int bodyW = 55;
        int bodyH = 25;
        int legW = 30;
        int legH = 10;
        int feetW = 6;
        int feetH = 6;
        int mouthW = 85;
        int mouthH = 10;
        int eyeW = 30;
        int eyeH = 25;
    
	// DRAW HIS RED HEAD
        gc.setFill(Paint.valueOf(headColor));
	gc.fillRect(location.getX(), location.getY(), headW, headH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX(), location.getY(), headW, headH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+32, location.getY()+88, bodyW, bodyH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+32, location.getY()+88, bodyW, bodyH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+45, location.getY()+112, legW, legH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+45, location.getY()+112, legW, legH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+40, location.getY()+122, feetW, feetH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+40, location.getY()+122, feetW, feetH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+75, location.getY()+122, feetW, feetH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+75, location.getY()+122, feetW, feetH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+15, location.getY()+60, mouthW, mouthH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+15, location.getY()+60, mouthW, mouthH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(eyeColor));
	gc.fillRect(location.getX()+15, location.getY()+15, eyeW, eyeH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+15, location.getY()+15, eyeW, eyeH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(eyeColor));
	gc.fillRect(location.getX()+70, location.getY()+15, eyeW, eyeH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+70, location.getY()+15, eyeW, eyeH);
	gc.stroke();
        
        gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+82, location.getY()+25, feetW, feetH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+82, location.getY()+25, feetW, feetH);
	gc.stroke();
        
         gc.setFill(Paint.valueOf(bodyColor));
	gc.fillRect(location.getX()+30, location.getY()+25, feetW, feetH);
        gc.beginPath();
	gc.setStroke(Paint.valueOf(outlineColor));
	gc.setLineWidth(1);
	gc.rect(location.getX()+30, location.getY()+25, feetW, feetH);
	gc.stroke();
        
    }

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
	launch(args);
    }
    
}
